bodo/transforms/distributed_analysis_call_registry.py

```diff
@@ -0,0 +1,634 @@
+            ): analyze_call_sklearn_metrics,
+            (
+                "confusion_matrix",
+                "sklearn.metrics._classification",
+            ): analyze_call_sklearn_metrics,
+            (
+                "confusion_matrix",
+                "sklearn.metrics",
+            ): analyze_call_sklearn_metrics,
+        }
+
+    def analyze_call(self, ctx, inst, fdef):
+        """Perform distributed analysis for input call if it's in the registry.
+        Return True if handled.
+
+        Args:
+            ctx (call_registry): distributed analysis context data
+            inst (ir.Assign): call expression
+            fdef (tuple[ir.Var|str]): function path in tuple of Var/string format
+
+        Returns:
+            bool: True if handled, False otherwise
+        """
+        handler_func = self.call_map.get(fdef, None)
+        if handler_func:
+            handler_func(ctx, inst)
+            return True
+
+        return False
+
+
+def no_op_analysis(ctx, inst):
+    """Handler that doesn't change any distributions"""
+    pass
+
+
+def meet_out_first_arg_analysis(ctx, inst):
+    """Handler that meets distributions of first call argument and output variables"""
+    _meet_array_dists(
+        ctx.typemap, inst.target.name, inst.value.args[0].name, ctx.array_dists
+    )
+
+
+def meet_first_2_args_analysis(ctx, inst):
+    """Handler that meets distributions of the first two call arguments"""
+    _meet_array_dists(
+        ctx.typemap, inst.value.args[0].name, inst.value.args[1].name, ctx.array_dists
+    )
+
+
+def meet_first_2_args_analysis_xgb_fit(ctx, inst):  # pragma: no cover
+    """Handler that meets distributions of the first two call arguments for xgb fit"""
+    _meet_array_dists(
+        ctx.typemap, inst.value.args[0].name, inst.value.args[1].name, ctx.array_dists
+    )
+    if ctx.array_dists[inst.value.args[0].name] == Distribution.REP:
+        raise BodoError("Arguments of xgboost.fit are not distributed", inst.loc)
+
+
+def analyze_mappable_table_funcs(ctx, inst):
+    """
+    Analyze for functions using generate_mappable_table_func.
+    Arg0 is a table used for distribution and arg1 is the function
+    name. Distributions differ based on arg1.
+
+    Returns True if there was a known implementation.
+    """
+    lhs = inst.target.name
+    rhs = inst.value
+    func_name_typ = ctx.typemap[rhs.args[1].name]
+    has_func = not is_overload_none(func_name_typ)
+    if has_func:
+        # XXX: Make this more scalable by recalling the distributed
+        # analysis already in this pass for each of the provided
+        # func names.
+        func_name = guard(get_overload_const_str, func_name_typ)
+        # We support mappable prefixes that don't need to be separate functions.
+        if func_name[0] == "~":
+            func_name = func_name[1:]
+        # Note: This isn't an elif because the ~ may modify the name
+        if func_name in (
+            "bodo.libs.array_ops.array_op_isna",
```

Comment:
We should find a way to move this to some more generic name registration, but this is not urgent at this time.